activation/nipost.go

```diff
@@ -377,64 +356,130 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 
 	round, err := client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID)
 	if err != nil {
-		return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		return nipost.PoETRegistration{},
+			&PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
 	}
 	logger.Info("challenge submitted to poet proving service", zap.String("round", round.ID))
-	return nipost.AddPoetRegistration(nb.localDB, nodeID, nipost.PoETRegistration{
+
+	registration := nipost.PoETRegistration{
 		ChallengeHash: types.Hash32(challenge),
 		Address:       client.Address(),
 		RoundID:       round.ID,
 		RoundEnd:      round.End,
-	})
+	}
+
+	if err := nipost.AddPoetRegistration(nb.localDB, nodeID, registration); err != nil {
+		return nipost.PoETRegistration{}, err
+	}
+
+	return registration, err
 }
 
-// Submit the challenge to all registered PoETs.
+// submitPoetChallenges submit the challenge to registered PoETs
+// if some registrations are missing and PoET round didn't start.
 func (nb *NIPostBuilder) submitPoetChallenges(
 	ctx context.Context,
 	signer *signing.EdSigner,
-	deadline time.Time,
+	poetProofDeadline time.Time,
+	curPoetRoundStartDeadline time.Time,
 	challenge []byte,
-) error {
+) ([]nipost.PoETRegistration, error) {
+	// check if some registrations missing or were removed
+	nodeID := signer.NodeID()
+	registrations, err := nipost.PoetRegistrations(nb.localDB, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get poet registrations from db: %w", err)
+	}
+
+	registrationsMap := make(map[string]nipost.PoETRegistration)
+	for _, reg := range registrations {
+		registrationsMap[reg.Address] = reg
+	}
+
+	existingRegistrationsMap := make(map[string]nipost.PoETRegistration)
+	var missingRegistrations []PoetService
+	for addr, poet := range nb.poetProvers {
+		if val, ok := registrationsMap[addr]; ok {
+			existingRegistrationsMap[addr] = val
+		} else {
+			missingRegistrations = append(missingRegistrations, poet)
+		}
+	}
+
+	existingRegistrations := maps.Values(existingRegistrationsMap)
+	if len(missingRegistrations) == 0 {
+		return existingRegistrations, nil
+	}
+
+	now := time.Now()
+
+	if curPoetRoundStartDeadline.Before(now) {
+		if len(existingRegistrations) == 0 {
+			if len(registrations) == 0 {
+				// no existing registration at all, drop current registration challenge
+				err = ErrATXChallengeExpired
+			} else {
+				// no existing registration for given poets set
+				err = ErrNoRegistrationForGivenPoetFound
+				nb.logger.Warn("revert poet configuration to previous is needed immediately",
+					zap.Error(err), log.ZShortStringer("smesherID", nodeID))
+			}
+			return nil, fmt.Errorf(
+				"%w: poet round has already started at %s (now: %s)",
+				err,
+				curPoetRoundStartDeadline,
+				now,
+			)
+		}
+		return existingRegistrations, nil
+	}
```

Comment:
NIT: this can be simplified

```suggestion
	if curPoetRoundStartDeadline.Before(now) {
		switch {
		case len(existingRegistrations) == 0 && len(registrations) == 0:
			// no existing registration at all, drop current registration challenge
			return nil, fmt.Errorf(
				"%w: poet round has already started at %s (now: %s)",
				ErrATXChallengeExpired,
				curPoetRoundStartDeadline,
				now,
			)
		case len(existingRegistrations) == 0:
			// TODO(mafa): In my opinion the node should exit here
			// no existing registration for given poets set
			nb.logger.Warn("revert poet configuration to previous is needed immediately",
				zap.Error(err), log.ZShortStringer("smesherID", nodeID))
			}
			return nil, fmt.Errorf(
				"%w: poet round has already started at %s (now: %s)",
				ErrNoRegistrationForGivenPoetFound,
				curPoetRoundStartDeadline,
				now,
			)
		default:
			return existingRegistrations, nil
		}
	}
```

Reply:
yes, but in my case I repeat less code, for example returning and error

Reply:
I see no repeated code in my suggested snipped? ❓

Reply:
Also with panic this would be: 

```go
	if curPoetRoundStartDeadline.Before(now) {
		switch {
		case len(existingRegistrations) == 0 && len(registrations) == 0:
			// no existing registration at all, drop current registration challenge
			return nil, fmt.Errorf(
				"%w: poet round has already started at %s (now: %s)",
				ErrATXChallengeExpired,
				curPoetRoundStartDeadline,
				now,
			)
		case len(existingRegistrations) == 0:
			// no existing registration for given poets set
			nb.logger.Panic(
				"Invalid PoET registrations found. Local state only contains PoETs not in node config",
				zap.Strings("registrations", maps.Keys(registrationsMap)),
				zap.Strings("configured_poets", maps.Keys(nb.poetProvers)),
				log.ZShortStringer("smesherID", nodeID),
			)
		default:
			return existingRegistrations, nil
		}
	}
```